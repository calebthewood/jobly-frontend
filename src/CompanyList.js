import JoblyApi from './api';
import React, { useState, useEffect } from "react";
import CompanyCard from './CompanyCard';
import SearchForm from './SearchForm';
import Loading from './Loading';

/** Manages a list of Companies
 *
 * state -> [{company}, {company}, ...]
 *
 * prop - none
 *
 * RouteList -> Route -> CompanyList
 */

function CompanyList() {
  const [companies, setCompanies] = useState([]);
  const [isLoading, setIsLoading] = useState(true);

  /**Gets list of companies on mount */
  useEffect(function getCompaniesFromApi() {
    async function getCompanies() {
      const companies = await JoblyApi.getCompanies();
      setCompanies(companies);
      setIsLoading(false);
    }
    getCompanies();
  }, []);

  /**Handles Search, updates companies state */
  async function search(term) {
    setIsLoading(true);
    const companies = await JoblyApi.searchCompany(term);
    setCompanies(companies);
    setIsLoading(false);
  }

  const notFound = <p style={{ color: "white" }}>Companies not found.</p>;



  if (isLoading) return <Loading />

  return (
    <div className="container">
      <div>
        <SearchForm search={search} />
      </div>

      {companies.length ?
        <div id="companyList">
          {companies.map(company => <CompanyCard key={company.handle} company={company} />)}
        </div>
        : notFound}
    </div>
  );
}

export default CompanyList;