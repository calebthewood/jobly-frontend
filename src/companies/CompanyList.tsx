import JoblyApi from '../api/api';
import { useState, useEffect } from "react";
import CompanyCard from './CompanyCard';
import SearchForm from '../common/SearchForm';
import Loading from '../common/Loading';
import { ICompany } from '../interfaces';

/** Manages a list of Companies
 *
 * state -> [{company}, {company}, ...]
 *
 * prop - none
 *
 * RouteList -> Route -> CompanyList
 */
function CompanyList() {
  const [companies, setCompanies] = useState<ICompany[] | []>([]);
  const [isLoading, setIsLoading] = useState(true);

  /**Gets list of companies on mount */
  useEffect(function getCompaniesFromApi() {
    async function getCompanies() {
      const companies: ICompany[]  = await JoblyApi.getCompanies();
      setCompanies(companies);
      setIsLoading(false);
    }
    getCompanies();
  }, []);

  /**Handles Search, updates companies state */
  async function search(term: string) {
    setIsLoading(true);
    const companies: ICompany[] = await JoblyApi.searchCompany(term);
    setCompanies(companies);
    setIsLoading(false);
  }

  const notFound = <p data-testid="not-found" style={{ color: "white" }}>Companies not found.</p>;

  if (isLoading) return <Loading />

  return (
    <div className="container">
      <div data-testid="search">
        <SearchForm search={search} />
      </div>

      {companies.length ?
        <div id="companyList" data-testid="resolved">
          {companies.map( (company: ICompany) => <CompanyCard key={company.handle} company={company} />)}
        </div>
        : notFound}
    </div>
  );
}

export default CompanyList;
