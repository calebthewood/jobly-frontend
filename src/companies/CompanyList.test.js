import { render, screen, waitFor } from '@testing-library/react';
import "@testing-library/jest-dom/extend-expect";
import { MemoryRouter } from "react-router-dom";
import { UserProvider } from "../testUtils";
import CompanyList from './CompanyList';
import JoblyApi from "../api/api";


jest.mock("../api/api");

const companies = [
  {
    handle: "test-company-1",
    name: "Test Company 1",
    description: "This is a test company.",
    numEmployees: 245,
    logoUrl: "/logos/logo3.png",
  }, {
    handle: "test-company-2",
    name: "Test Company 2",
    description: "This is another test company.",
    numEmployees: 10,
    logoUrl: null,
  }
];

describe('CompanyList', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('matches snapshot', function () {
    const { asFragment } = render(<CompanyList />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders Loading...', async () => {
    render(<CompanyList />);
    expect(screen.getByTestId("loading")).toBeInTheDocument();
    expect(screen.getByTestId("loading")).toHaveTextContent("Loading...");
  });

  it('should render multiple companies', async () => {
    JoblyApi.getCompanies.mockResolvedValueOnce(companies);
    render(
      <MemoryRouter initialEntries={['/companies']}>
        <UserProvider>
          <CompanyList />
        </UserProvider>
      </MemoryRouter >
    );
    expect(screen.getByTestId("loading")).toBeInTheDocument();
    expect(screen.getByTestId("loading")).toHaveTextContent("Loading...");
    await waitFor(() => { expect(JoblyApi.getCompanies).toHaveBeenCalledTimes(1); });
    await waitFor(() => { expect(screen.getByText(companies[0].name)).toBeInTheDocument(); });
    await waitFor(() => { expect(screen.getByText(companies[0].description)).toBeInTheDocument(); });
    await waitFor(() => { expect(screen.getByText(companies[1].name)).toBeInTheDocument(); });
    await waitFor(() => { expect(screen.getByText(companies[1].description)).toBeInTheDocument(); });

  });

  it('should render "not found" message if state empty', async () => {
    JoblyApi.getCompanies.mockResolvedValueOnce([]);
    render(
      <MemoryRouter initialEntries={['/companies']}>
        <UserProvider>
          <CompanyList />
        </UserProvider>
      </MemoryRouter >
    );
    await waitFor(() => { expect(JoblyApi.getCompanies).toHaveBeenCalledTimes(1); });
    await waitFor(() => { expect(screen.getByTestId('not-found')).toBeInTheDocument(); });
    await waitFor(() => { expect(screen.getByText('Companies not found.')).toBeInTheDocument(); });
  });

  it('renders the search bar', async () => {
    JoblyApi.getCompanies.mockResolvedValueOnce(companies);
    render(
      <MemoryRouter initialEntries={['/companies']}>
        <UserProvider>
          <CompanyList />
        </UserProvider>
      </MemoryRouter >
    );
    await waitFor(() => { expect(screen.getByTestId('search')).toBeInTheDocument(); });
  });
});


