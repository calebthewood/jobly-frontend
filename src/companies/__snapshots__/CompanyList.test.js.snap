// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CompanyList matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="home-heading row align-items-center"
  >
    <div
      class="col"
    >
      <h2
        class="h2 fw-bold"
        data-testid="loading"
      >
        <i>
          Loading...
        </i>
      </h2>
    </div>
  </div>
</DocumentFragment>
`;
